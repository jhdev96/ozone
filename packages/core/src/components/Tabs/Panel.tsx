import React from "react";
import styles from './index.module.scss';


interface TabPanel {
  /**
   * The panel content.
   */
  children?: React.ReactNode;
}

export default function TabPanel({children}: TabPanel): JSX.Element 
{
  return (
    <div className={styles.panel}>
      {children}
    </div>
  )
}

TabPanel.displayName = 'TabPanel';