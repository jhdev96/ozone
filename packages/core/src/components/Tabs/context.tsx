import React from 'react';


interface ContextType {
  /** The name of the active tab. */
  activeTab: string;
  /** A function that sets the active tab. */
  setActiveTab: (name: string) => void;
};

export const TabContext = React.createContext<ContextType>({
  activeTab: '',
  setActiveTab: (name: string) => {}
});