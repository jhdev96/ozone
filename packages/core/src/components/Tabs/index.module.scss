@import '../../scss/root';


.tabs {
  display: inline-flex;

  .tab {
    width: fit-content;
    padding: 0.25rem 1rem;
    border-radius: 20px;
    cursor: pointer;
  }
}

.panel { margin: 10px 0; }

// context
.active {
  background-color: var(--blue-2);
}

.light {
  color: var(--black-core);
}

.dark:not(.active) {
  color: var(--gray-1);
}