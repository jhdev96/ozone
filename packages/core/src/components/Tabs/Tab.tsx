import React, { useContext } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import { TabContext } from './context';
import styles from './index.module.scss';


interface TabProps {
  /**
   * A required title for the tab. 
   * 
   * When a tab is clicked, its title
   * will be reported back to the 
   * `Tabs` component.
   */
  title: string;
  /**
   * Gives the tab a light or dark theme.
   */
  theme?: ThemeType;
  /**
   * Applies custom css styling to 
   * the component.
   */
  className?: string;
}

const Tab = React.forwardRef<HTMLDivElement, TabProps>(
  function Tab({
    title,
    theme = 'light',
    className}: TabProps,
    ref: React.ForwardedRef<HTMLDivElement>): JSX.Element 
  {
    const {activeTab, setActiveTab} = useContext(TabContext);

    const handleClick = () => {
      if (activeTab !== title) {
        setActiveTab(title);
      }
    }

    return (
      <div 
        className={classNames(
          className,
          styles.tab,
          styles[theme],
          activeTab === title && styles.active
        )}
        onClick={handleClick}
        ref={ref}
      >
        {title}
      </div>
    )
  }
)

Tab.displayName = 'Tab';

export default Tab;