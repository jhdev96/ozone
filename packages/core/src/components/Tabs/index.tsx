import React, { Component } from 'react';
import { TabContext } from './context';
import Tab from './Tab';
import TabPanel from './Panel';
import styles from './index.module.scss';


interface TabsProps {
  /**
   * The name of the selected tab.
   */
  selectedTab: string;
  /** 
   * The function that is called when
   * the the active tab is changed.
   */ 
  onChange?: (tabName?: string) => void;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /**
   * The individual tabs renderd inside 
   * the component.
   */
  children?: React.ReactNode;
}

type TabsState = {activeTab: string};

export default class Tabs extends Component<TabsProps, TabsState> {
  public static displayName = 'Tabs';

  public readonly state = {
    activeTab: this.props.selectedTab
  }

  componentDidUpdate() {
    if (this.props.onChange) {
      this.props.onChange(this.state.activeTab);
    }
  }

  setActiveTab = (tab: string) => {
    this.setState({activeTab: tab});
  }

  public render(): JSX.Element {
    const {activeTab} = this.state;
    const {children, dataTestId} = this.props;

    return (
      <TabContext.Provider 
        value={{
          activeTab, 
          setActiveTab: this.setActiveTab
        }}
      >
        <div 
          className={styles.tabs} 
          data-testid={dataTestId}
        >
          {children}
        </div>
      </TabContext.Provider>
    )
  }
}

export { Tab, TabPanel };