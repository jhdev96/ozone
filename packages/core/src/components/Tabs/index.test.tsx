import React from "react";
import { 
  render, 
  screen, 
  fireEvent 
} from "@testing-library/react";
import Tabs, { Tab, TabPanel } from '.';


describe('<Tabs />', () => {
  it('renders 3 tabs', () => {
    render(
      <Tabs dataTestId="test" selectedTab="apple">
        <Tab title="apple" />
        <Tab title="banana" />
        <Tab title="orange" />
      </Tabs>
    );

    expect(screen.getByTestId('test').children)
      .toHaveLength(3);
  });

  it('renders 3 tabs and a panel', () => {
    const {container} = render(
      <>
        <Tabs selectedTab="apple">
          <Tab title="apple" />
          <Tab title="banana" />
          <Tab title="orange" />
        </Tabs>
        <TabPanel>
          This is a panel
        </TabPanel>
      </>
    );

    expect(container).toMatchSnapshot();
  });

  it('selects the first tab', () => {
    render(
      <Tabs selectedTab="apple" dataTestId="test">
        <Tab title="apple" />
        <Tab title="banana" />
        <Tab title="orange" />
      </Tabs>
    );

    expect(screen.getByTestId('test').firstChild)
      .toHaveClass('active');
  });

  it('changes the selected tab when a new tab is clicked', () => {
    render(
      <Tabs selectedTab="apple" dataTestId="test">
        <Tab title="apple" />
        <Tab title="banana" />
        <Tab title="orange" />
      </Tabs>
    );

    const firstTab = screen.getByTestId('test').firstChild;
    const secondTab = screen.getByTestId('test').children[1];

    fireEvent.click(secondTab);

    expect(secondTab).toHaveClass('active');
    expect(firstTab).not.toHaveClass('active');
  });

  it('calls `onChange` when a new tab is selected', () => {
    const onChange = jest.fn();

    render(
      <Tabs 
        selectedTab="apple" 
        onChange={onChange}
        dataTestId="test"
      >
        <Tab title="apple" />
        <Tab title="banana" />
        <Tab title="orange" />
      </Tabs>
    );

    const secondTab = screen.getByTestId('test').children[1];
    fireEvent.click(secondTab);

    expect(onChange).toBeCalled();
  });
});