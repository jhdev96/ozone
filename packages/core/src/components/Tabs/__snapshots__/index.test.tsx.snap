// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> renders 3 tabs and a panel 1`] = `
<div>
  <div
    class="tabs"
  >
    <div
      class="tab light active"
    >
      apple
    </div>
    <div
      class="tab light"
    >
      banana
    </div>
    <div
      class="tab light"
    >
      orange
    </div>
  </div>
  <div
    class="panel"
  >
    This is a panel
  </div>
</div>
`;
